### ISSUE TITLE
Slow processing

### ISSUE THREAD

**Issue by jeroenhendricksen** [RaimondB/Detecticam#8]
**Problem description summary**

I'd like to report a bug that caused processing to be delayed with a few minutes. When the detect-i-cam process cannot keep up, it does detect motion, but this processing is delayed.

**Problem description details**

I have a second process running, which uses another method do determine that there is motion (tensorflow running in a docker container). It uses image snapshots, but that's beyond the point.

Detection for camera's using either method (a) image captures analyzed via tensorflow and b) detect-i-cam )are sent to Elasticsearch (after being published on MQTT, they end up in OpenHAB, after which I use Node-RED to store these events in Elasticsearch, every second; this is an near-realtime process), and the screenshot below shows any logged events using Kibana:

<img width="1128" alt="Screen Shot 2020-07-27 at 14 00 43" src="https://user-images.githubusercontent.com/32140294/88539614-ded8e800-d011-11ea-8aa7-e929eab86c00.png">

This is what the item names originate from:
- `Detecticam_CamFrontdoor_Motion` --> Motion from my frontdoor camera, as determined by detect-i-cam
- `CamFrontdoor_Motion` --> Motion from my frontdoor as detected by tensorflow
- etc...

It clearly shows that detect-i-cam observations are processed minutes after they actually happen, which should have been almost immediate. What's positive about this on a sidenote is that they both agree on there being true (what I call) "human motion" and I can confirm that there are no worrisome false positives in the shown data sample.

**What I expect**
- That the data being processed from the 3 RTSP streams at 2 FPS kan be kept up with and if not:
  - Images that are not process within the next 10 sec (or so) should be ignore, only to retry with the latest image.
  - The log should indicate that the process cannot keep up (if that's the case)

**This is my config**

```
{
    "video-streams": [{
        "id": "FRONTDOOR",
        "path": "rtsp://",
        "fps": 2,
        "callbackUrl": ""
    },{
        "id": "DRIVEWAY",
        "path": "rtsp://",
        "fps": 2,
        "callbackUrl": ""
    },{
    	"id": "BARN",
        "path": "rtsp://",
        "fps": 2,
        "callbackUrl": ""
    }],
    "mqtt-publisher": {
        "enabled": true,
        "server": "rabbitmq",
        "port": 1883,
        "topicPrefix": "detecticam",
        "clientId": "detecticam"
    }
}
```

CPU usage looks like this:
<img width="896" alt="Screen Shot 2020-07-27 at 13 42 52" src="https://user-images.githubusercontent.com/32140294/88538106-172af700-d00f-11ea-8ee1-b9ea12459794.png">

Memory usage looks like this:
<img width="755" alt="Screen Shot 2020-07-27 at 13 43 12" src="https://user-images.githubusercontent.com/32140294/88538136-227e2280-d00f-11ea-8a1f-a22772559532.png">

Logs from the container look like this: [docker.log.txt](https://drive.google.com/file/d/1QY2vaEauF74sC2tD4BLfgm3VJkchwAEO/view?usp=sharing)

Please let me know if you need more information. It seems I can easily reproduce this. I have not yet played with an increase or reduction in FPS, lowering the number of observed videostreams, or looking at the effect of a container restart.

**Comment by RaimondB** [comment]
Is this on the dev tag or the latest tag?

**Comment by jeroenhendricksen** [comment]
Forgot to mention that. It's running the latest dev tag.

Op ma 27 jul. 2020 19:22 schreef Raimond Brookman <notifications@github.com

**Comment by RaimondB** [comment]
Btw I think I know whats happening here. There is also a related issue, that causes the batches to be out of sync. E.g. first camera image might be timed at x, and from a second camera stream at x+3 sec. This is because each videostream capture keeps its own time on when to publish the next frame for analysis. This needs to be externalized, so that, even when errors occur there is a steady 'tick' in parallel for all streams in sync.

I did have in an earlier version a timeout. But that is a bit tricky. The neural network I am using is running as a singleton and to save memory allocations, the memory of the analysis output is reused. This means I cannot have running analysis concurrently. Because they will be writing in the same output area.
Also, when I did the timeout, this would only cancel the managed part, since the neural network in OpenCV is on a native thread and cannot easily be cancelled.
So I can put out warnings that analysis cannot keep up, but it is a bit hard to abort running analysis at the moment.

Btw I am not having this problem, but this might be because I am not getting any decoding errors. I see a lot of restarts of capture etc. Usually this means you are having to much CPU pressure, because of which capture cannot keep up. When you are missing frames, the decoder runs into issues. And then things get worse from there.

I'll take a look at producing the 'steady tick'  and warning about timeouts.

**Comment by jeroenhendricksen** [comment]
> Btw I am not having this problem, but this might be because I am not getting any decoding errors. I see a lot of restarts of capture etc. Usually this means you are having to much CPU pressure, because of which capture cannot keep up. When you are missing frames, the decoder runs into issues. And then things get worse from there.
> 

I've reduced the configuration to a single camera, removed the fps configuration and restarted the container. Now it seems I no longer have any decoding errors. CPU usage is now averaging at around 12%. When I look at events, they seem to occur quickly and immediately. I will look if it still performs detection immediately after the container has been running for a while.

<img width="1209" alt="Screen Shot 2020-07-29 at 21 58 19" src="https://user-images.githubusercontent.com/32140294/88847336-2a90ba80-d1e7-11ea-9e5a-d0b7de5c3720.png">

**Comment by jeroenhendricksen** [comment]
> Btw I think I know whats happening here. There is also a related issue, that causes the batches to be out of sync. E.g. first camera image might be timed at x, and from a second camera stream at x+3 sec. This is because each videostream capture keeps its own time on when to publish the next frame for analysis. This needs to be externalized, so that, even when errors occur there is a steady 'tick' in parallel for all streams in sync.
> 
> I did have in an earlier version a timeout. But that is a bit tricky. The neural network I am using is running as a singleton and to save memory allocations, the memory of the analysis output is reused. This means I cannot have running analysis concurrently. Because they will be writing in the same output area.
> Also, when I did the timeout, this would only cancel the managed part, since the neural network in OpenCV is on a native thread and cannot easily be cancelled.
> So I can put out warnings that analysis cannot keep up, but it is a bit hard to abort running analysis at the moment.

I have an easy solution/workaround/hack for this issue. I can easily start a separate container, so multiple videostreams are processed by their own native thread. Of course this would probably not benefit overall performance and processing efficiency. What are your thoughts on this hack, would you recommend it?

**Comment by RaimondB** [comment]
I built the batching approach to save on memory and CPU. CPU for 2x single camera will Be more then for 2 cameras batched because of batching efficiency of the dnn. Also the YOLO network takes up a lot of mem that you will need twice. When you don't have issues with the resources indeed you could configure multiple containers. But it does not scale well when you e.g have 8 cameras.

**Comment by RaimondB** [comment]
Btw, it should also be possible to configure the other videostreams as well. The only thing you need to take into account is that the classifyobjects ms in the logs stay below 3000, to prevent it from lagging.

**Comment by RaimondB** [comment]
@jeroenhendricksen I am now testing out the new dev tag that makes sure camera feeds are triggered at the same time, and also processed in sync. You can try it if you want.

**Comment by jeroenhendricksen** [comment]
> @jeroenhendricksen I am now testing out the new dev tag that makes sure camera feeds are triggered at the same time, and also processed in sync. You can try it if you want.

It's working without lag now, running for 2 camera's. 

CPU averages around 22% and the memory usage is stable (700MB)
 
<img width="944" alt="Screen Shot 2020-08-04 at 19 33 06" src="https://user-images.githubusercontent.com/32140294/89325628-5b5e6d00-d689-11ea-8e30-0513e8a7c25f.png">

Detection is now much faster (not lagging behind), even after it has been running for a while:

<img width="460" alt="Screen Shot 2020-08-04 at 19 36 39" src="https://user-images.githubusercontent.com/32140294/89326075-08d18080-d68a-11ea-8c6c-c3058e0c9413.png">

<img width="468" alt="Screen Shot 2020-08-04 at 19 36 23" src="https://user-images.githubusercontent.com/32140294/89326081-0bcc7100-d68a-11ea-8e6d-1093351ecf9b.png">

Another issue that now remains is the detection being too "sensitive", compared to the tensorflow analysis. That will be "fixed" for me when certainty percentages for detected objects are also published to MQTT.

Thx for fixing!
